### DISCUSSION TITLE
Vue + TSX examples

### DISCUSSION THREAD

**Discussion by matthew-dean** [storybookjs/storybook#26879] [Help] [answered]
### Summary

How do you use `.tsx` stories with Vue? I know JSX/TSX are supported in Vue in general, but I can't seem to get any of my Vue stories to render when it's a `stories.tsx` file, and none of the Storybook blog posts / guides for Vue mention this approach.

**Comment by matthew-dean** [comment, marked as answer]
Had to scour the internet but I found the solution!

First, you have to make sure that tsconfig.json has this:

```json
  "compilerOptions": {
     "jsx": "preserve",
    "jsxImportSource": "vue"
  }
```

Then, in the Vite config, it needs this:
```js
  esbuild: {
    jsxFactory: 'h',
    jsxFragment: 'Fragment',
    jsxInject: `import { h, Fragment } from 'vue'`
  },
```

Then, in the `*.stories.tsx` files, it looks like:
```tsx
export const Default: Story = {
  render: () => <Button />
}
```

**Reply by ahmedragab20** [reply]
the problem with this solution is that when I implement it, the components is no longer reactive with the control panel (when toggling a boolean prop it doesn't listen to the change for example).. have you had the same issue?

**Reply by matthew-dean** [reply]
@ahmedragab20 Try something like this:

```tsx
export const Default: Story = {
  args: {
    modelValue: false,
  },
  render: (args) => ({
    setup() {
      return { args };
    },
    render: () => <Checkbox {...args} />,
  }),
};
```

What's a little counter-intuitive is that the render in the StoryObj is Storybook's render, not Vue's render. So you need to return a render-able component. That can be a function OR can be a whole Vue component definition. And since we're using TSX, then you need another render method in the returned object.

Note that this will not propagate upwards. That is, checking the checkbox in this example will not update the state in your control panel.
